File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/Handlers/DiscountHandler.cs
namespace ChainOfResponsibilityDesignPattern.Example1.Handlers;

public class DiscountHandler : BaseOrderHandler
{
    private readonly Dictionary<string, decimal> _discountCodes;

    public DiscountHandler()
    {
        _discountCodes = new Dictionary<string, decimal>
        {
            { "SUMMER20", 0.20m },
            { "WELCOME10", 0.10m },
            { "VIP30", 0.30m },
            { "NEWYEAR15", 0.15m }
        };
    }

    public override void Handle(OrderRequest request)
    {
        if (string.IsNullOrEmpty(request.DiscountCode))
        {
            request.AddMessage("No discount code applied");
            base.Handle(request);
            return;
        }

        if (_discountCodes.ContainsKey(request.DiscountCode))
        {
            var discountPercentage = _discountCodes[request.DiscountCode];
            var discountAmount = request.TotalAmount * discountPercentage;
            request.TotalAmount -= discountAmount;
            
            request.AddMessage($"Discount applied: {discountPercentage * 100}% off, saved ${discountAmount:F2}");
        }
        else
        {
            request.AddMessage($"Invalid discount code: '{request.DiscountCode}'");
        }
        
        base.Handle(request);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/src/Handlers/DiscountHandler.ts
import { BaseOrderHandler } from "../BaseOrderHandler";
import { OrderRequest } from "../OrderRequest";

export class DiscountHandler extends BaseOrderHandler {
	private readonly _discountCodes: Map<string, number> = new Map([
		["SUMMER20",   0.20],
		["WELCOME10",  0.10],
		["VIP30",      0.30],
		["NEWYEAR15",  0.15],
	]);

	handle(request: OrderRequest): void {
		if (!request.discountCode) {
			request.addMessage("No discount code applied");
			super.handle(request);
			return;
		}
		const rate = this._discountCodes.get(request.discountCode);
		if (rate !== undefined) {
			const discountAmount = request.totalAmount * rate;
			request.totalAmount -= discountAmount;
			request.addMessage(`Discount applied: ${rate * 100}% off, saved $${discountAmount.toFixed(2)}`);
		} else {
			request.addMessage(`Invalid discount code: '${request.discountCode}'`);
		}
		super.handle(request);
	}
}
